### ISSUE TITLE
[PROPOSAL] TagsFacet for arbitrary metadata

### ISSUE THREAD

**Issue by leogodin217** [OpenLineage/OpenLineage#3169]
**Purpose:**

Arbitrary metadata, often referred to as tagging, concerns adding additional datapoints that are not part of the OL spec to events. 

Several use cases exist for this type of data. 

- Environment information for jobs
- Setting PII/security information for columns
- Grain for datasets
- Technology-specific attributes i.e. (dbt Cloud project, Airflow tags, etc.)
- Downstream notification information like Slack channels, email addresses, etc.
- Configuration information for downstream automation (A message to re-run a pipeline, create Jira ticket, kick off reverse ETL, etc.)
- Enriching events with domain-specific tags
- etc.

No matter how many fields OpenLineage adds to the spec, there will always be a need for additional metadata both supplied by the client and the user. In order to allow OpenLineage to support more use cases, we should consider a generalized and automatable approach to capturing this type of metadata. 

*See these discussions for additional context:*

https://github.com/OpenLineage/OpenLineage/issues/2779
https://github.com/OpenLineage/OpenLineage/issues/2748
https://github.com/OpenLineage/OpenLineage/pull/1630

*Examples*

Arbitrary metadata comes in many forms. Some implementations assign single-string tokens as tags. Others allow key/value pairs or even full dictionaries. Our solution needs to handle all of these cases. Furthermore, the metadata can come from integrations, users or the specific tools we are monitoring with OpenLineage. It would be good to track the provenience of each one. 

Airflow DAG (Single token strings)

```python
    @dag(
        start_date=pendulum.datetime(2024, 6, 10, tz="UTC"),
        tags=["crr_pipeline", "prod", "team=dataos"],
        schedule="5,15,25,35,45,55 * * * *", 
    )
```

dbt meta (Dictionary of arbitrary metadata)

```yaml
   - name: subscription_rate_plan_view
      config:
        meta:
          nr_config:
            team: Order Management 
            alert_failed_test_rows: True 
            failed_test_rows_limit: 20 
            slack_mentions: '@user1, @user2, @user3
```

Snowflake column tags (Schema-level object stored as a key/value pair)

```sql
    ALTER TABLE hr.tables.empl_info
    MODIFY COLUMN job_title
    SET TAG visibility = 'public';
```

User-supplied tag through environment variables

```bash
    export OPENLINEAGE_TAG_ENVIRONMENT=prod
```

Run Event Example with Tags
```json
{
        "eventType": "START",
        "eventTime": "2020-12-28T19:52:00.001+10:00",
        "run": {
          "runId": "0176a8c2-fe01-7439-87e6-56a1a1b4029f",
          "tags": [
            {"key": "project", "value": "myproject", "source": "DBT_CLOUD_INTEGRATION"},
            {"key": "environment", "value": "production", "source": "DBT_CLOUD_INTEGRATION"},
            {"key": "crr_pipeline", "value": true, "source": "USER"},
            {"key": "is_adhoc", "value": true, "source": "DBT_CLOUD_INTEGRATION"}
            
          ]
        },
        "job": {
          "namespace": "my-namespace",
          "name": "my-job",
          "facets": {
              "tags": [
                {"key": "project", "value": "myproject", "source": "DBT_CLOUD_INTEGRATION"},
                {"key": "environment", "value": "production", "source": "DBT_CLOUD_INTEGRATION"},
                {"key": "crr_pipeline", "value": true, "source": "USER"}
              ]
          }
        },
        "inputs": [{
          "namespace": "my-namespace",
          "name": "my-input",
          "facets": {
            "tags": [
                {"key": "meta_nr_config_slack_mentions": "value": "@user1, @user2", "source": "DBT_CLOUD_INTEGRATION"},
                {"key": "pii", "value": true, "field": "email", "source": "DBT_CLOUD_INTEGRATION"}
            ],
            "schema": {
                "fields": [
                    {"name": "email", "type": "string" },
                    {"name": "id", "type": "int"}
                ] 
                
            }
          }
        }],  
        "producer": "https://github.com/OpenLineage/OpenLineage/blob/v1-0-0/client",
        "schemaURL": "https://openlineage.io/spec/1-0-5/OpenLineage.json#/definitions/RunEvent"
      }
```
**Proposed implementation**

This implementation proposes a new TagFacet that can be applied to any job, run, or dataset field. The facet will have one property called tags. Tags is an array of objects holding key/value pairs and optional identifying information. While using key/value does not preserve the hierarchical nature of nested objects, it does allow us to collect all needed information supporting the vast majority of use cases.

This schema will include a TagFacet with four properties. 
* key: Identifies the tag by name
* value: The value of the tag. Set to true to mimic single-string token. (pii=true)
* source: (optional) Where the tag came from. This can be from the integration, a user, or the technology we are monitoring. 
* field: (optional) Dataset field the tag applies to

To add context to tags, we will generate TagJobFacet, TagRunFacet and TagDatasetFacet all referencing TagFacet.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://openlineage.io/spec/facets/1-1-1/TagFacet.json",
  "$defs": {
    "TagFacetFields": {
      "type": "object",
      "properties": {
        "key": {
          "description": "Key that identifies the tag",
          "type": "string",
          "example": "pii"
        },
        "value": {
          "description": "The value of the field",
          "type": ["string", "number", "boolean"],
          "example": "true|@user1|production"
        },
        "source": {
          "description": "The source of the tag. INTEGRATION|USER|DBT CORE|SPARK|etc.",
          "type": "string"
        },
        "field": {
          "description": "Identifies the field in a dataset if a tag applies to one",
          "type": "string",
          "example": "email_address"
        },
      },
      "required": ["key", "value"]
    },
    "TagFacet": {
      "allOf": [
        {
          "$ref": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/BaseFacet"
        },
        {
          "type": "object",
          "properties": {
            "tags": {
              "description": "The tags applied to the facet",
              "type": "array",
              "items": {
                "$ref": "#/$defs/TagFacetFields"
              }
            }
          }
        }
      ],
      "type": "object"
  },
  "TagRunFacet": {
      "allOf": [
        {
          "$ref": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/TagFacet"
        },
        {
          "type": "object",
          "properties": {}
          }
      ],
      "type": "object"
  },
  "TagJobFacet": {
      "allOf": [
        {
          "$ref": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/TagFacet"
        },
        {
          "type": "object",
          "properties": {}
        }
      ],
      "type": "object"
  },
  "TagDatasetFacet": {
      "allOf": [
        {
          "$ref": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/TagFacet"
        },
        {
          "type": "object",
          "properties": {}
        }
      ],
      "type": "object"
  },
}
```

Tag Lifecycle

Tags can be created, modified and deleted. However, it is difficult for stateless integrations to identify changes. The
integrations generally only understand current state. To enable deletion and modification, events should always  
provide the full set of tags. Consumers should consider tags from the latest event to be the complete set of tags. 

Python and Java client

* Implement the tag facet
* Identify user-supplied tags through environment variables i.e. (OPENLINEAGE_TAG__ENVIRONMENT=production)
* Allow integrations to query for user-supplied tags

Integrations

* Identify technology-specific tags and metadata, create the tag facet, attach the tag facet to appropriate parent facets
* Optional - Allow user-supplied tags through configuration or CLI arguments i.e. (--environment=production)

Building Tags

* The combination of key and source should be unique within an event.
* Single-token tags use key=True. i.e. (pii becomes {key: pii, value: true}) 
* Key should match the source data point as closely as possible. i.e. ("team=dataos", "pii" and "production" become key=team, key=pii, key=production)
* User-supplied tags supplied through environment variables or config files should apply to job and run, but not dataset. (Dataset should have a method of supplying tags through an integration)
* In cases where the source metadata is an object, key/value pairs should be built from the object. For example, the dictionary below would be represented as a single tag. 

```yaml
   - name: subscription_rate_plan_view
      config:
        meta:
          nr_config:
            team: Order Management 
```

becomes 

```json
  {
    "key": "meta_nr_config_team",
    "value": "Order Management"
  }
```

**Comment by mobuchowski** [comment]
@leogodin217 +1 - the proposal looks really good, and addresses what we've been discussing - such as provenience of the tags. 


One nit I would have is
```
OPENLINEAGE_TAG_ENVIRONMENT=prod -> OPENLINEAGE_TAG__KEY=prod
```
- easier to distinguish between "env name" part and "tag key".

**Comment by JDarDagran** [comment]
@leogodin217 Thank you for your excellent proposal. I have a few suggestions to further clarify and enhance it:

* **Example Events with New Facet:**
Please provide concrete examples of events that would benefit from the addition of a new "tags" facet. This will help us visualize the practical applications and potential value of this feature.

* **Tag Configuration:**
I propose that tags be primarily configured by the user through existing mechanisms like configuration files or environment variables. This aligns with our current approach for other configuration options and avoids introducing a new exception. This flexibility will also accommodate different use cases, such as dynamic tag updates based on runtime conditions.

* **Tag Placement:**
While integrations may attach the "tags" facet to various parent facets, clients need clear guidance on where to add specific tags. Consider these options:

  * **New Property**: Introduce a dedicated property within the relevant event type to hold the "tags" facet. For example, in the RunEvent type, we could add a new optional tags property. This would provide a well-defined location for clients to add tags. For instance, `RunEvent` now expects several fields like `run`, `job`, `schemaURL` etc. but none of these is purely a `tag` field.
  * **Convention-Based Placement**: Establish a convention or mapping that determines the appropriate parent facet for different types of tags. This could be based on tag categories or event attributes.

**Comment by leogodin217** [comment]
@mobuchowski 

> One nit I would have is
> 
> OPENLINEAGE_TAG_ENVIRONMENT=prod -> OPENLINEAGE_TAG__KEY=prod
> easier to distinguish between "env name" part and "tag key".

I'm not sure I understand. My thought is it should be OPENLINEAGE_TAG__key=value. Assign both the key and value in the same variable. Are you suggesting single-token tags that do not have a key? Something that would render like:

```json
    tags: [prod, pii, ...]
```

instead of 

```json
    tags: [{env=prod}, {pii=true}, ...]
```

**Comment by leogodin217** [comment]
@JDarDagran 

> Example Events with New Facet:
I'll update it tomorrow

>Tag Configuration:
Many tags come from the technology itself. Furthermore, those tags provide context to what event it should apply to. I'm not sure how a user could configure all that outside of the the technology. For instance, a tag on an Airflow task. How would that be manually configured outside the DAG code? 

> Tag Placement:
I was envisioning this to be the integration responsibility. I could see the client maybe attaching user-supplied tags to jobs and runs, but the integration would also have that responsibility. (Maybe I lack understanding of how the events are created)

> New Property:
We discussed this in Slack and the google doc. Seems like the consensus was that tags belong in custom facets. 

>Convention-Based Placement: 
Hmm, I don't think we are proposing meaning behind specific tags. That being said, we should consider examples like Airflow. Do DAG tags apply to the dataset? Or just the job and run? That's where it gets difficult to come up with global conventions. Do you have any suggestions?

**Comment by JDarDagran** [comment]
> Many tags come from the technology itself. Furthermore, those tags provide context to what event it should apply to. I'm not sure how a user could configure all that outside of the the technology. For instance, a tag on an Airflow task. How would that be manually configured outside the DAG code?

I can see two ways of configuring it then. One is identical to other configurations we currently have in clients, like transport (setting URLs etc.). For instance, you'd want to attach instance ID on which Airflow runs - this could be just set in configuration file / dynamic env var. Another way is to let integrations interact with clients and pass tags to them (that's at least how I see it, correct me if the proposal states different approach).



> Convention-Based Placement:
> Hmm, I don't think we are proposing meaning behind specific tags. That being said, we should consider examples like Airflow. Do DAG tags apply to the dataset? Or just the job and run? That's where it gets difficult to come up with global conventions. Do you have any suggestions?

I didn't mean proposing meaning behind specific tags. In Google Doc I can see some discussion around having e.g. `TagDatasetFacet`, `TagJobFacet`. This way tag facets would be put under their parent facets (e.g. `TagDatasetFacet` goes with dataset facets).

The difficulty of differentiating if something applies to job/run/dataset stays the same regardless of discussion about tags.

**Comment by leogodin217** [comment]
@JDarDagran 

> can see two ways of configuring it then. One is identical to other configurations we currently have in clients, like transport >(setting URLs etc.). For instance, you'd want to attach instance ID on which Airflow runs - this could be just set in >configuration file / dynamic env var. Another way is to let integrations interact with clients and pass tags to them (that's at >least how I see it, correct me if the proposal states different approach).
Yes, I'm proposing both methods. User supplied tags identified by the client through config/variables and integration tags automatically pulled from the tools. In both cases I think the integration should attach the tag to the appropriate facet. 

For tool supplied tags (Airflow tags, dbt meta, etc....) integration identifies the tag -> integration attaches tag to correct facets. 

For user supplied tags (config, environment variables) client identifies tag -> integration queries the client -> integration attaches tag to correct facets.

 >I didn't mean proposing meaning behind specific tags. In Google Doc I can see some discussion around having e.g. >TagDatasetFacet, TagJobFacet. This way tag facets would be put under their parent facets (e.g. TagDatasetFacet goes >with dataset facets).
I don't have a strong preference either way. Whatever the community wants. @julienledem any thoughts on how to decide this? It seems like an important discussion.

**Comment by JDarDagran** [comment]
> Yes, I'm proposing both methods. User supplied tags identified by the client through config/variables and integration tags automatically pulled from the tools. In both cases I think the integration should attach the tag to the appropriate facet.

I just remembered about https://github.com/OpenLineage/OpenLineage/issues/3144#issuecomment-2404565636 proposal. If this gets implemented too, I think we could have mechanism for ruser supplied tags only. For tool supplied it could be just event mutation hook.

**Comment by JDarDagran** [comment]
> I don't have a strong preference either way. Whatever the community wants. @julienledem any thoughts on how to decide this? It seems like an important discussion.

I think some example could also help here. Those don't have to be full events.

**Comment by mobuchowski** [comment]
@leogodin217 

>I'm not sure I understand. My thought is it should be OPENLINEAGE_TAG__key=value. Assign both the key and value in the same variable. Are you suggesting single-token tags that do not have a key? Something that would render like:

No, that was just a nit about formatting of the env variable name.

**Comment by leogodin217** [comment]
@JDarDagran I updated the proposal with an example event. As for guidance, I think I misunderstood what you were asking for. You were looking for technical guidance, right? My proposal is more at the spec level then the technical implementation. Unfortunately, I don't have enough knowledge of the client internals to make a recommendation there. Maybe we can all crowdsource this.

**Comment by JDarDagran** [comment]
@leogodin217 thanks for the update! I don't expect you to delve into the technical implementation, this is just a discussion in the wider community and is something to keep in mind.

Looking at the example and based on the current solution, in my opinion we should add 3 facets (and probably one `TagFacet` serving as a parent for them with all the properties to inherit) - `TagJobFacet`, `TagRunFacet`, `TagDatasetFacet`. Again, this is probably more of an implementation detail.

In conclusion - I love the idea and I'm in favor of implementing this. That's a well written proposal, thanks again!

**Comment by leogodin217** [comment]
@JDarDagran I added the additional facets. While doing that, tag mutations came to mind. It seems like it would be very difficult for integrations to identify changes in tags. Is the best way to handle this to consider the tags on the latest event as the complete, accurate set? This is what I wrote. 

> Tags can be created, modified and deleted. However, it is difficult for stateless integrations to identify changes. The
> integrations generally only understand current state. To enable deletion and modification, events should always
> provide the full set of tags. Consumers should consider tags from the latest event to be the complete set of tags.
